extralibs/rtime/rtime.man
.TH rtime 3x "3 Jul 1995" " "
.SH NAME
clock_settime, clock_gettime, clock_getres \- high-resolution clock
.SH SYNOPSIS
.LP
.ti -2n
PARIX:
.br
.B px ancc
.B \-I/usr/local/include
.RI "[ " "flag" " \|.\|.\|. ] " "file" " \|.\|.\|."
.B "\-L/usr/local/lib \-lrtime"
.RI "[ " "library" " \|.\|.\|. ]"
.LP
.ti -2n
Solaris:
.br
.B cc 
.B \-I/usr/local/include
.RI "[ " "flag" " \|.\|.\|. ] " "file" " \|.\|.\|."
.B "\-lposix4 \-L/usr/local/lib \-lrtime"
.RI "[ " "library" " \|.\|.\|. ]"
.LP
.ti -2n
AIX:
.br
.B xlc
.B \-I/usr/local/include
.RI "[ " "flag" " \|.\|.\|. ] " "file" " \|.\|.\|."
.B "\-L/usr/local/lib \-lrtime"
.RI "[ " "library" " \|.\|.\|. ]"
.LP
.ti -2n
OSF/1:
.br
.B cc 
.B \-I/usr/local/include
.RI "[ " "flag" " \|.\|.\|. ] " "file" " \|.\|.\|."
.B "\-lrt \-L/usr/local/lib \-lrtime"
.RI "[ " "library" " \|.\|.\|. ]"
.LP
.ti -2n
---
.br
.B
#include "rtime.h"
.LP
.BI "int clock_settime(clockid_t " "clock_id, "
.if n .ti +5n
.BI "const struct timespec *" "tp" ");"
.LP
.BI "int clock_gettime(clockid_t " "clock_id" ", struct timespec *" "tp" ");"
.LP
.BI "int clock_getres(clockid_t " "clock_id" ", struct timespec *" "res" ");"
.LP
.ta +5n +14n +16n
.B
struct  timespec {
.br
.B
	time_t	tv_sec;	\f1/* seconds */\fP
.br
.B
	long	tv_nsec;	\f1/* and nanoseconds */\fP
.br
.B
};

.SH DESCRIPTION
.IX "clock_settime" "" "\fLclock_settime\fP \(em high-resolution clock operations" ""
.IX "clock_gettime" "" "\fLclock_gettime\fP \(em high-resolution clock operations" ""
.IX "clock_getres" "" "\fLclock_getres\fP \(em high-resolution clock operations" ""
.B clock_settime(\|)
sets the specified clock,
.I clock_id,
to the value specified by
.I tp.
.LP
.B clock_gettime(\|)
returns the current value
.I tp
for the specified clock,
.I clock_id.
.LP
The resolution of any clock can be obtained by calling
.B clock_getres(\|).
If
.I res
is not NULL,
the resolution of the specified clock is stored
in
.I res.
.LP
The
.I clock_id
for the real-time clock for the system is \f4CLOCK_REALTIME\fP.
If 
.B clock_settime(\|)
has not been invoked, 
.B clock_gettime(\|)
returns the amount of time (in seconds and nanoseconds) since the
application started.
.SH "RETURN VALUES"
.B clock_settime(\|), clock_gettime(\|),
and
.B clock_getres(\|)
return
.B 0
upon success, otherwise they return
.B -1
and set 
.B errno
to indicate the error condition.
.SH ERRORS
.TP 12
.B EINVAL
.I clock_id
does not specify a known clock.
.br
- 
The 
.I tp
argument to
.B clock_settime(\|)
is outside the range for the given clock id.
.br
-
The
.I tp
argument to 
.B clock_settime(\|)
specified a nanosecond
value less than zero or greater than or equal to 1,000,000,000.
.TP
.B ENOSYS
.B clock_settime(\|),
.B clock_gettime(\|),
or
.B clock_getres(\|)
is not supported by this implementation.
.TP
.B EPERM
The requesting process does not have the appropriate privilege
to set the specified clock.
.SH NOTES
Clock resolutions are implementation defined and are not
settable.
Time values that are between two consecutive non-negative integer
multiples of the resolution of the specified clock are truncated
down to the smaller multiple of the resolution.

\f4CLOCK_REALTIME\fP is the only implemented clock so far \|.\|.\|.

This library works currently on 
.B PARIX 1.3(.1) 
(only PowerPC601),
.B Solaris 2.4, 
.B AIX 3.2, 
and 
.B OSF/1 V3.2.

In
.B PARIX,
any user can set the clock with
.B clock_settime(\|)
and is only valid as long as current application run.

.SH AUTHOR
Niclas Andersson <nican@ida.liu.se>
.br 
Department of Computer and Information Science
.br 
Linkoping University, Sweden


extralibs/rtime/rtime.c
/* 

   This package uses the RTCU and RTCL registers in PowerPC 601 to meassure
   time. Note: Other PowerPC's don't have these registers.
   
   The resolution of the RTCU is 1 second and of the RTCL is 128
   nanoseconds.
   
   It wraps around after 2^32 seconds = ca. 136 years.
   
   
   Safe 

   The PARIX-1.3 call TimeNow() also use the information in these
   registers but it truncates the result to microseconds. Also, since
   the result is returned in a unsigned int, it wraps around after
   ca. 71 minutes.
*/

#include "rtime.h"

#ifdef __PARIX__

#include <errno.h>
#include <sys/sem.h>

static Semaphore_t sem = {1, 0, 0};
static volatile timespec_t Diff;

#endif

/* ---------------------------------------------------------------- */

void timespecadd(timespec_t *sum, timespec_t *t1, timespec_t *t2)
{
    sum->tv_nsec = t1->tv_nsec + t2->tv_nsec;
    sum->tv_sec  = t1->tv_sec  + t2->tv_sec;
    if (sum->tv_nsec > NANOSEC) {
	sum->tv_nsec -= NANOSEC;
	sum->tv_sec ++;
    }
}

void timespecdiff(timespec_t *diff, timespec_t *t1, timespec_t *t2)
{
    diff->tv_nsec = t1->tv_nsec - t2->tv_nsec;
    diff->tv_sec  = t1->tv_sec  - t2->tv_sec;
    if (diff->tv_nsec < 0) {
	diff->tv_nsec += NANOSEC;
	diff->tv_sec --;
    }
}

#ifdef __PARIX__
int clock_settime(clockid_t clock_id, timespec_t *tp)
{
    timespec_t diff, now;

    if (clock_id != CLOCK_REALTIME ||
	tp->tv_nsec < 0 || tp->tv_nsec >= NANOSEC) {
	errno = EINVAL;
	return -1;
    }
    rtctime(&now);
    timespecdiff(&diff, tp, &now);
    Wait(&sem);
    Diff = diff;
    Signal(&sem);

    return 0;
}

int clock_gettime(clockid_t clock_id, timespec_t *tp)
{
    timespec_t now, diff;

    if (clock_id != CLOCK_REALTIME) {
	errno = EINVAL;
	return -1;
    }
    Wait(&sem);
    diff = Diff;
    Signal(&sem);
    rtctime(&now);
    timespecadd(tp, &diff, &now);
    
    return 0;
}

int clock_getres(clockid_t clock_id, timespec_t *res)
{
    if (clock_id != CLOCK_REALTIME) {
	errno = EINVAL;
	return -1;
    }
    res->tv_sec = 0;
    res->tv_nsec = 128;
    
    return 0;
}

#endif


#ifdef _UNICOS
/** 
     Added Patrik Nordling in the year of the horse......
     Based on gettimeofday(2)
 **/

#include <errno.h>

int clock_settime(clockid_t clock_id, timespec_t *tp)
{
    struct timeval t ;
    if (clock_id != CLOCK_REALTIME ||
	tp->tv_nsec < 0 || tp->tv_nsec >= NANOSEC) {
        errno = EINVAL;
	return -1;
    }
    if(gettimeofday(&t, NULL)){
        perror("gettimeofday failed") ; }
    tp->tv_sec = t.tv_sec ;
    tp->tv_nsec = t.tv_usec*1000 ;
    return 0;
}

int clock_gettime(clockid_t clock_id, timespec_t *tp)
{
    struct timeval t ;
    if (clock_id != CLOCK_REALTIME) {
	errno = EINVAL;
	return -1;
    }
    if(gettimeofday(&t, NULL)){
        perror("gettimeofday failed") ; }    
    tp->tv_sec = t.tv_sec ;
    tp->tv_nsec = t.tv_usec*1000 ;
    return 0;
}

int clock_getres(clockid_t clock_id, timespec_t *res)
{
    if (clock_id != CLOCK_REALTIME) {
	errno = EINVAL;
	return -1;
    }
    res->tv_sec = 0;
    res->tv_nsec = 128;
    return 0;
}

#endif

/*#if defined (__linux__) || defined (__CYGWIN__)*/
/* Alex: Linux-2.2 implementes clock_settime in time.h*/
/* Hakan: But need to link -lrt  */

#if defined (__CYGWIN__)

#include <errno.h>
#include <stdlib.h>

int clock_settime(clockid_t clock_id, timespec_t *tp)
{
    struct timeval t ;
    if (clock_id != CLOCK_REALTIME ||
	tp->tv_nsec < 0 || tp->tv_nsec >= NANOSEC) {
        errno = EINVAL;
	return -1;
    }
    if(gettimeofday(&t, NULL)){
        perror("gettimeofday failed") ; }
    tp->tv_sec = t.tv_sec ;
    tp->tv_nsec = t.tv_usec*1000 ;
    return 0;
}

int clock_gettime(clockid_t clock_id, timespec_t *tp)
{
    struct timeval t ;
    if (clock_id != CLOCK_REALTIME) {
	errno = EINVAL;
	return -1;
    }
    if(gettimeofday(&t, NULL)){
        perror("gettimeofday failed") ; }    
    tp->tv_sec = t.tv_sec ;
    tp->tv_nsec = t.tv_usec*1000 ;
    return 0;
}

int clock_getres(clockid_t clock_id, timespec_t *res)
{
    if (clock_id != CLOCK_REALTIME) {
	errno = EINVAL;
	return -1;
    }
    res->tv_sec = 0;
    res->tv_nsec = 128;
    return 0;
}

#endif


#if defined (_MSC_VER) || defined(__MINGW32__)

/* 
   New approach for MSC using clock()
   clock() returns the proces time 
   since process start.
   A.Siemers 00.01.25
 */

#include <time.h>
#include <winsock.h>
#include <errno.h>

int clock_settime(clockid_t clock_id, timespec_t *tp)
{
    clock_t secs;

    if (clock_id != CLOCK_REALTIME ||
	tp->tv_nsec < 0 || tp->tv_nsec >= NANOSEC) {
        errno = EINVAL;
	return -1;
    }

    secs = clock();

//    secs = secs / CLOCKS_PER_SEC;
//    tp->tv_sec = secs;
//    tp->tv_nsec = secs*1000;

    tp->tv_sec = secs / CLOCKS_PER_SEC;
    tp->tv_nsec = (secs % CLOCKS_PER_SEC)*(NANOSEC/CLOCKS_PER_SEC);


    return 0;
}

int clock_gettime(clockid_t clock_id, timespec_t *tp)
{
    clock_t secs;

    if (clock_id != CLOCK_REALTIME) {
	errno = EINVAL;
	return -1;
    }

    secs = clock();
//    secs = secs / CLOCKS_PER_SEC;
//    tp->tv_sec = secs;
//    tp->tv_nsec = secs*1000;

    tp->tv_sec = secs / CLOCKS_PER_SEC;
    tp->tv_nsec = (secs % CLOCKS_PER_SEC)*(NANOSEC/CLOCKS_PER_SEC);

    return 0;
}

int clock_getres(clockid_t clock_id, timespec_t *res)
{
    if (clock_id != CLOCK_REALTIME) {
	errno = EINVAL;
	return -1;
    }
    res->tv_sec = 0;
//    res->tv_nsec = 128;
    res->tv_nsec= (NANOSEC/CLOCKS_PER_SEC);
    return 0;
}

#endif
